src/Communicate/BufferHandler.hpp

```diff
@@ -0,0 +1,142 @@
+#ifndef IPPL_BUFFER_HANDLER_HPP
+#define IPPL_BUFFER_HANDLER_HPP
+
+
+template <typename MemorySpace>
+BufferHandler<MemorySpace>::~BufferHandler() {}
+
+template <typename MemorySpace>
+typename BufferHandler<MemorySpace>::buffer_type 
+BufferHandler<MemorySpace>::getBuffer(size_type size, double overallocation) {
+    size_type requiredSize = static_cast<size_type>(size * overallocation);
+
+    auto freeBuffer = findFreeBuffer(requiredSize);
+    if (freeBuffer != nullptr) {
+        return allocateFromFreeBuffer(freeBuffer);
+    }
+
+    if (!free_buffers.empty()) {
+        return reallocateLargestFreeBuffer(requiredSize);
+    }
+
+    return allocateNewBuffer(requiredSize);
+}
+
+template <typename MemorySpace>
+void BufferHandler<MemorySpace>::freeBuffer(buffer_type buffer) {
+    if (isBufferUsed(buffer)) {
+        releaseUsedBuffer(buffer);
+    }
+}
+
+template <typename MemorySpace>
+void BufferHandler<MemorySpace>::freeAllBuffers() {
+    free_buffers.insert(used_buffers.begin(), used_buffers.end());
+    used_buffers.clear();
+
+    freeSize += allocatedSize;
+    allocatedSize = 0;
+}
+
+template <typename MemorySpace>
+void BufferHandler<MemorySpace>::deleteAllBuffers() {
+    freeSize      = 0;
+    allocatedSize = 0;
+
+    used_buffers.clear();
+    free_buffers.clear();
+}
+
+template <typename MemorySpace>
+typename BufferHandler<MemorySpace>::size_type 
+BufferHandler<MemorySpace>::getAllocatedSize() const {
+    return allocatedSize;
+}
+
+template <typename MemorySpace>
+typename BufferHandler<MemorySpace>::size_type 
+BufferHandler<MemorySpace>::getFreeSize() const {
+    return freeSize;
+}
+
+template <typename MemorySpace>
+bool BufferHandler<MemorySpace>::bufferSizeComparator(const buffer_type& lhs, const buffer_type& rhs) {
+    if (lhs->getBufferSize() != rhs->getBufferSize()) {
+        return lhs->getBufferSize() < rhs->getBufferSize();
+    }
+    return lhs < rhs;
```

Comment:
In order to find an element based on size, the set requires a total ordering among the elements and this return is a tie-breaker based on the (unique) address, therefore enforcing a total order among the buffers.

Probably it's a good idea to document this with a comment in the code.

Reply:
Yes, I noticed this size ordering, but is the size ordering the same as the address ordering. I am not sure this is guaranteed by `std::set`, is it?

Reply:
Wait, I don't quite understand your question. So I have a bunch of objects that need to be sorted. If i just compared them using their size attribute, it would not result in a total ordering as two buffers with the same size would be ordered either one way or the other. In case they have the same size, I look at the memory address of the buffer (the thing I give out to customers) that is clearly a unique identifier for the buffers. There is no guarantees used from `std::set`.

Reply:
OK. Out of curiosity, do we need a total ordering? I think it should not matter if, for example, buffer A and buffer B with size(A) = size(B) are ordered A, B or B, A. I am fine with the current solution, however.

Reply:
Yes it is required that the comparison function is a strict weak ordering.
According to the cppreference of [`std::set`](https://en.cppreference.com/w/cpp/container/set), it follows the requirements of [AssociativeContainer](https://en.cppreference.com/w/cpp/named_req/AssociativeContainer) which has the requirement that the elements must follow a strict weak ordering. And a strict weak ordering is asymmetric so `a < b` and `b < a` can not both be true at the same time.

Reply:
OK. Thanks!